cmd/tempo-serverless/handler.go

```diff
@@ -0,0 +1,208 @@
+package handler
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/gogo/protobuf/jsonpb"
+	"github.com/grafana/tempo/pkg/api"
+	"github.com/grafana/tempo/pkg/model"
+	"github.com/grafana/tempo/pkg/tempopb"
+	"github.com/grafana/tempo/tempodb"
+	"github.com/grafana/tempo/tempodb/backend"
+	"github.com/grafana/tempo/tempodb/backend/azure"
+	"github.com/grafana/tempo/tempodb/backend/gcs"
+	"github.com/grafana/tempo/tempodb/backend/local"
+	"github.com/grafana/tempo/tempodb/backend/s3"
+	"github.com/grafana/tempo/tempodb/encoding"
+	"github.com/mitchellh/mapstructure"
+	"github.com/spf13/viper"
+	"gopkg.in/yaml.v2"
+
+	// required by the goog
+	_ "github.com/GoogleCloudPlatform/functions-framework-go/funcframework"
+)
+
+const envConfigPrefix = "TEMPO"
+
+// todo(search)
+//   integration tests
+//   pass meta.json as a body here and to the queriers
+
+// Handler is the main entrypoint
+// Parameters
+//  - searchRequest           - tags, min/maxDuration, limit
+//  - BackendSearch 		  - start, end
+//  - BackendSearchQuerier    - startPage, totalPages, blockID
+//  - BackendSearchServerless - encoding, dataEncoding, indexPageSize, totalRecords, tenant, version
+// Response
+//  - tempopb.SearchResponse
+func Handler(w http.ResponseWriter, r *http.Request) {
+	searchReq, err := api.ParseSearchRequest(r, 20, 0) // todo(search): parameterize 20
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	// parseSearchRequest doesn't respect these as "reserved" tags. let's remove them here.
+	// this will all be cleaned up when search paths are consolidated.
+	delete(searchReq.Tags, api.URLParamBlockID)
+	delete(searchReq.Tags, api.URLParamStartPage)
+	delete(searchReq.Tags, api.URLParamTotalPages)
+	delete(searchReq.Tags, api.URLParamStart)
+	delete(searchReq.Tags, api.URLParamEnd)
+	delete(searchReq.Tags, api.URLParamEncoding)
+	delete(searchReq.Tags, api.URLParamIndexPageSize)
+	delete(searchReq.Tags, api.URLParamTotalRecords)
+	delete(searchReq.Tags, api.URLParamTenant)
+	delete(searchReq.Tags, api.URLParamDataEncoding)
+	delete(searchReq.Tags, api.URLParamVersion)
+
+	start, end, limit, err := api.ParseBackendSearch(r)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	startPage, totalPages, blockID, err := api.ParseBackendSearchQuerier(r)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	enc, dataEncoding, indexPageSize, totalRecords, tenant, version, err := api.ParseBackendSearchServerless(r)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	// load local backend config file. by convention this must be in the same folder at ./config.json
+	reader, err := loadBackend()
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
```

Comment:
Something to look into later: this code is not specific to the incoming request, would it make sense to move this into some kind of init function? I'm not sure how Google Functions work, but with Lambda's you can run expensive initialisation code once and reuse it for multiple invocations.

Reply:
Dang. Good catch. Used `sync.Once()` to prevent reinitialization.